packages/snaps-controllers/src/multichain/MultichainRouter.ts

```diff
@@ -0,0 +1,387 @@
+};
+
+type ProtocolSnap = {
+  snapId: SnapId;
+  methods: string[];
+};
+
+const name = 'MultichainRouter';
+
+export class MultichainRouter {
+  #messenger: MultichainRouterMessenger;
+
+  #withSnapKeyring: WithSnapKeyringFunction;
+
+  constructor({ messenger, withSnapKeyring }: MultichainRouterArgs) {
+    this.#messenger = messenger;
+    this.#withSnapKeyring = withSnapKeyring;
+
+    this.#messenger.registerActionHandler(
+      `${name}:handleRequest`,
+      async (...args) => this.handleRequest(...args),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${name}:getSupportedMethods`,
+      (...args) => this.getSupportedMethods(...args),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${name}:getSupportedAccounts`,
+      (...args) => this.getSupportedAccounts(...args),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${name}:isSupportedScope`,
+      (...args) => this.isSupportedScope(...args),
+    );
+  }
+
+  /**
+   * Attempts to resolve the account address to use for a given request by inspecting the request itself.
+   *
+   * The request is sent to to an account Snap via the SnapKeyring that will attempt this resolution.
+   *
+   * @param snapId - The ID of the Snap to send the request to.
+   * @param scope - The CAIP-2 scope for the request.
+   * @param request - The JSON-RPC request.
+   * @returns The resolved address if found, otherwise null.
+   * @throws If the invocation of the SnapKeyring fails.
+   */
+  async #resolveRequestAddress(
+    snapId: SnapId,
+    scope: CaipChainId,
+    request: JsonRpcRequest,
+  ) {
+    try {
+      const result = (await this.#withSnapKeyring(async (keyring) =>
+        keyring.resolveAccountAddress(snapId, scope, request),
+      )) as { address: CaipAccountId } | null;
+      const address = result?.address;
+      return address ? parseCaipAccountId(address).address : null;
+    } catch {
+      throw rpcErrors.internal();
+    }
+  }
+
+  /**
+   * Get the account ID of the account that should service the RPC request via an account Snap.
+   *
+   * This function checks whether any accounts exist that can service a given request by
+   * using a combination of the resolveAccountAddress functionality and the connected accounts.
+   *
+   * If an account is expected to service this request but none is found, the function will throw.
+   *
+   * @param connectedAddresses - The CAIP-10 addresses connected to the requesting origin.
+   * @param scope - The CAIP-2 scope for the request.
+   * @param request - The JSON-RPC request.
+   * @returns An account ID if found, otherwise null.
+   * @throws If no account is found, but the accounts exist that could service the request.
+   */
+  async #getSnapAccountId(
+    connectedAddresses: CaipAccountId[],
+    scope: CaipChainId,
+    request: JsonRpcRequest,
+  ) {
+    const accounts = this.#messenger
+      .call('AccountsController:listMultichainAccounts', scope)
```

Comment:
Note for us @danroc.

This method relies on a old logic (prior to the `scopes` introduction), and we should definitely update this now, see:
- https://github.com/MetaMask/core/blob/v290.0.0/packages/accounts-controller/src/AccountsController.ts#L1044-L1046

Reply:
Definitely agreed. The router assumes this call uses the scopes as much as possible 😄

Reply:
I suppose it's time to do that `TODO` :dadjoke:

Reply:
- https://github.com/MetaMask/accounts-planning/issues/820